src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,168 @@
+   */
+  private class ScopeTreeGenerator {
+
+    /* Keep track of the # of created Scatter nodes. */
+    var scatterCount: Int = 0
+
+  /*
+  * Retrieve the list of AstNodes from the AST body.
+  * Return empty seq if body is null or is not a list.
+  */
+    private def getChildrenList(ast: Ast): Seq[AstNode] = {
+      val body: AstNode = ast.getAttribute("body")
+      if(body == null) Seq.empty
+      else {
+        body match {
+          case list: AstList => list.asScala.toVector
+          case _ => Seq.empty
+        }
+      }
+    }
+
+    /**
+     * Generate recursively all descendant Scopes for the workflow.
+     * Should only be called during apply with a root Ast
+     * @param node current working node
+     * @param namespaces namespaces
+     * @param tasks tasks
+     * @param wdlSyntaxErrorFormatter formatter
+     * @return Some(instance of Scope) if the node has a Scope equivalent, None otherwise
+     */
+    def generateScopeTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Option[Scope] = {
+      node match {
+          //WORKFLOW
+        case w: Ast if w.getName.equals(AstNodeName.Workflow) =>
+          val workflowChildren = getChildrenList(w) flatMap (generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter))
+          val workflow: Workflow = Workflow(w, wdlSyntaxErrorFormatter)
+          workflow.setChildren(workflowChildren)
+          Some(workflow)
+
+          //CALL
+        case c: Ast if c.getName.equals(AstNodeName.Call) => Some(Call(c, namespaces, tasks, wdlSyntaxErrorFormatter))
+
+          //SCATTER
+        case s: Ast if s.getName.equals(AstNodeName.Scatter) =>
+          // Need to create the Scatter before generating its children to have the correct index layout order
+          val scatter = Scatter(s, scatterCount)
+          scatterCount += 1
+          val scatterChildren = getChildrenList(s) flatMap (generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter))
+          scatter.setChildren(scatterChildren)
+          Some(scatter)
+
+        /* Cases need to be added here to support new scopes */
+        case _ => None
+      }
+    }
+  }
+
+  /**
+   * Generate a Scope Tree from an Ast Node.
+   * Node must match a Scope (currently: Workflow, Scatter or Scope)
+   * @param node node to build tree from
+   * @param namespaces namespaces used
+   * @param tasks tasks
+   * @param wdlSyntaxErrorFormatter wdl formatter
+   * @throws UnsupportedOperationException if node does not match any supported Scopes
+   * @return generated Scope Tree.
+   */
+  def generateTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Scope = {
+    new ScopeTreeGenerator().generateScopeTree(node, namespaces, tasks, wdlSyntaxErrorFormatter)
+    .getOrElse(throw new UnsupportedOperationException("Input node cannot be instantiated as a Scope"))
```

Comment:
Yes I'm not fond of it either. However I tried something like you're mentioning but I wasn't quite able to make it work in all cases (nested scatters + siblings) to have it respect the "scatter index ordering" we want. 
Also I'm not sure having it return the scatterIndex is a good long term approach because if we add later some other type of Scope it's going to become quite complex to keep track of which index is returned, incremented etc... Here we can just add a new var attribute and use that for the new Scope.
In the end we just need a counter to know how much scatters we created so far, and the simplest way I could find to do that is to have a mutable "counter" variable. I'll keep looking for a way to do something as simple without var but so far every attempt made things more complicated.